Data_Engine/Query/DepthDictionary.cs

```diff
@@ -0,0 +1,81 @@
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.oM.Data.Collections;
+using System.Collections.Generic;
+using System.ComponentModel;
+using BH.oM.Reflection.Attributes;
+using System.Linq;
+using System;
+
+namespace BH.Engine.Data
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+        [Description("Gets the depth dictionary of a graph using breadth first search, each key value pair in the resulting dictionary is in the form <graph node, depth>")]
+        [Input("adjacency", "The adjacency dictionary to extract the depth dictionary from")]
+        [Input("startNode", "The graph node from which the depth dictionary is created")]
+        public static Dictionary<GraphNode<T>, int> DepthDictionary<T>(Dictionary<GraphNode<T>, List<GraphNode<T>>> adjacency, GraphNode<T> startNode)
+        {
+            //https://www.geeksforgeeks.org/level-node-tree-source-node-using-bfs/
+            if(!adjacency.ContainsKey(startNode))
+                throw new ArgumentException("startNode provided cannot be found in the adjacency dictionary. Ensure the node is from the original graph");
+            // dictionary to store level of each node  
+            Dictionary<GraphNode<T>, int> level = new Dictionary<GraphNode<T>, int>();
+            // dictionary to store when node has been visited
+            Dictionary<GraphNode<T>, bool> marked = new Dictionary<GraphNode<T>, bool>();
+            // create a queue  
+            Queue<GraphNode<T>> que = new Queue<GraphNode<T>>();
+            // enqueue element x  
+            que.Enqueue(startNode);
+            // initialize level of source node to 0  
+            level[startNode] = 0;
+            // marked it as visited  
+            marked[startNode] = true;
+            // do until queue is empty  
+            while (que.Count > 0)
+            {
+                // dequeue element  
+                startNode = que.Dequeue();
+                // traverse neighbors of node x  
+                foreach (GraphNode<T> b in adjacency[startNode])
```

Comment:
Since adjacency created outside of this method, it would be good to make sure that `startNode` is a key of `adjacency`.

Reply:
yes already using `if(!adjacency.ContainsKey(startNode))` on line 43. throwing ArgumentException but changing to `Reflection.Compute.RecordError` in line with other comments

Reply:
ha yes, true